rasa/model.py

```diff
@@ -340,42 +375,60 @@ def merge_model(source: Text, target: Text) -> bool:
         shutil.move(source, target)
         return True
     except Exception as e:
-        logging.debug(e)
+        logging.debug(f"Could not merge model: {e}")
         return False
 
 
-def should_retrain(new_fingerprint: Fingerprint, old_model: Text, train_path: Text):
-    """Checks which component of a model should be retrained.
+def should_retrain(
+    new_fingerprint: Fingerprint, old_model: Text, train_path: Text
+) -> FingerprintComparisonResult:
+    """Check which components of a model should be retrained.
 
     Args:
         new_fingerprint: The fingerprint of the new model to be trained.
         old_model: Path to the old zipped model file.
         train_path: Path to the directory in which the new model will be trained.
 
     Returns:
-        A tuple of boolean values indicating whether Rasa Core and/or Rasa NLU needs
+        A ShouldRetrain object indicating whether Rasa Core and/or Rasa NLU needs
         to be retrained or not.
 
     """
-    retrain_nlu = retrain_core = True
+    retrain = FingerprintComparisonResult()
 
     if old_model is None or not os.path.exists(old_model):
-        return retrain_core, retrain_nlu
+        return retrain
 
     with unpack_model(old_model) as unpacked:
         last_fingerprint = fingerprint_from_path(unpacked)
-
         old_core, old_nlu = get_model_subdirectories(unpacked)
 
-        if not core_fingerprint_changed(last_fingerprint, new_fingerprint):
-            target_path = os.path.join(train_path, "core")
-            retrain_core = not merge_model(old_core, target_path)
+        retrain = FingerprintComparisonResult(
+            core=did_section_fingerprint_change(
+                last_fingerprint, new_fingerprint, SECTION_CORE
+            ),
+            nlu=did_section_fingerprint_change(
+                last_fingerprint, new_fingerprint, SECTION_NLU
+            ),
+            nlg=did_section_fingerprint_change(
+                last_fingerprint, new_fingerprint, SECTION_NLG
+            ),
+        )
+
+        core_merge_failed = False
+        if not retrain.core:
+            target_path = os.path.join(train_path, DEFAULT_CORE_SUBDIRECTORY_NAME)
+            core_merge_failed = not merge_model(old_core, target_path)
+            retrain.core = core_merge_failed
+
+        if not retrain.nlg:
+            retrain.nlg = core_merge_failed
```

Comment:
Can you add a comment here why we overwrite `retrain.nlg` with `core_merge_failed`?

Reply:
also changed the code to make it a bit easier to read